cirq-core/cirq/sim/classical_simulator.py

```diff
@@ -12,96 +12,211 @@
+            )
+        elif isinstance(initial_state, List):
+            state = ClassicalBasisState(initial_state)
+        else:
+            raise ValueError('initial_state must be an int or List[int]')
+        super().__init__(state=state, qubits=qubits, classical_data=classical_data)
+
+    def _act_on_fallback_(self, action, qubits: Sequence['cirq.Qid'], allow_decompose: bool = True):
+        """Acts on the state with a given operation.
+
+        Args:
+            action: The operation to apply.
+            qubits: The qubits to apply the operation to.
+            allow_decompose: Whether to allow decomposition of the operation.
+
+        Returns:
+            True if the operation was applied successfully.
+
+        Raises:
+            ValueError: If gate is not one of X, CNOT, SWAP, CCNOT, or a measurement.
+        """
+        gate = action.gate if isinstance(action, ops.Operation) else action
+        mapped_qubits = [self.qubit_map[i] for i in qubits]
+        if _is_identity(gate):
+            return True
+        if gate == ops.X:
+            (q,) = mapped_qubits
+            self._state.basis[q] ^= 1
+            return True
+        elif gate == ops.CNOT:
+            c, q = mapped_qubits
+            self._state.basis[q] ^= self._state.basis[c]
+            return True
+        elif gate == ops.SWAP:
+            a, b = mapped_qubits
+            self._state.basis[a], self._state.basis[b] = self._state.basis[b], self._state.basis[a]
+            return True
+        elif gate == ops.TOFFOLI:
+            c1, c2, q = mapped_qubits
+            self._state.basis[q] ^= self._state.basis[c1] & self._state.basis[c2]
+            return True
+        else:
+            raise ValueError(f'{gate} is not one of X, CNOT, SWAP, CCNOT, or a measurement')
+
+
+class ClassicalStateStepResult(
+    sim.StepResultBase['ClassicalBasisSimState'], Generic[TSimulationState]
+):
+    """The step result provided by `ClassicalStateSimulator.simulate_moment_steps`."""
+
+
+class ClassicalStateTrialResult(
+    sim.SimulationTrialResultBase['ClassicalBasisSimState'], Generic[TSimulationState]
+):
+    """The trial result provided by `ClassicalStateSimulator.simulate`."""
+
+
+class ClassicalStateSimulator(
+    sim.SimulatorBase[
+        ClassicalStateStepResult['ClassicalBasisSimState'],
+        ClassicalStateTrialResult['ClassicalBasisSimState'],
+        'ClassicalBasisSimState',
+    ],
+    Generic[TSimulationState],
+):
+    """A simulator that accepts only gates with classical counterparts."""
+
+    def __init__(
+        self, *, noise: 'cirq.NOISE_MODEL_LIKE' = None, split_untangled_states: bool = False
+    ):
+        """Initializes a ClassicalStateSimulator.
+
+        Args:
+            noise: The noise model used by the simulator.
+            split_untangled_states: Whether to run the simulation as a product state.
+        """
+        super().__init__(noise=noise, split_untangled_states=split_untangled_states)
```

Comment:
general noise doesn't make sense here, so for now lets fail if a noise model is provided until we agree on which noise models to support.

```py3
if noise is not None: raise ValueError(f'{noise=} is not supported')
```